### DISCUSSION TITLE
Pure virtual method called

### DISCUSSION THREAD

**Discussion by shibin2018** [sparcians/map#472] [General]
When I terminate the simulation, it will throw an exception about pure virtual method called. I find it in ResourceTreeNode.hpp:132 when the deleteParameters function called with gdb.

**Comment by klingaard** [comment]
This is probably happening because a `ResourceFactory` is being destroyed _before_ the `ResourceTreeNode` that uses it is destroyed.  The order of destruction is important.  See below...

Below is an example of the order of operations with respect to factories and tree nodes.  (Pulled from the [overview presentation](https://docs.google.com/presentation/d/e/2PACX-1vS1BWtVv0x3qXKQWAeECe2gsF9cMG3Zp2HnXJw52grCAcl21lv3a9pLW6J0lZ32e5DWdZkFyUMcE_AI/pub?start=false&loop=false&delayms=3000), slide 13).
```
class CPU
{
private:
    // The factory MUST BE deleted AFTER the resource tree node.  The order is important here
    sparta::ResourceFactory<core_example::Core, core_example::Core::CoreParameterSet> core_rf;
    std::unique_ptr<sparta::ResourceTreeNode> core_ptr;
};

CPU::CPU(...)
{
     core_ptr.reset(new sparta::ResourceTreeNode(top, “core”, &core_rf));
}
```
An example can be found [here](https://github.com/riscv-software-src/riscv-perf-model/blob/6f2cbb777970b00fbe16ece8e0da9e31ac837915/core/Execute.hpp#L56-L64).

**Reply by shibin2018** [reply]
btw, in the ResourceTreeNode.hpp:63, the Constructor 
ResourceTreeNode(TreeNode* parent,
                         const std::string & name,
                         const std::string & group,
                         group_idx_type group_idx,
                         const std::string & desc,
                         ResourceFactoryBase* res_fact) :
            TreeNode(name, group, group_idx, desc),
            created_resource_(false),
            res_fact_(res_fact),
            params_(0),
            adata_(this, ARCH_DATA_LINE_SIZE))
this res_fact_(res_fact) use the raw pointer paramater res_fact initilal the member raw pointer variable res_fact_, here use the shared_ptr may be better?

**Reply by klingaard** [reply]
That would have been the smart thing to do.  :grin:   I'll look into doing that.

**Comment by shibin2018** [comment]
Thank you for reply!
but in my cases, I did not destroy the factory and treenode explicitly, Only in my top model, I use the getRoot()->enterTeardown() to free the resource automatically in my Destructor function.
by the way, has the future presentation mentioned in the last slice[overview presentation]  been done? I want to learn more about the sparta with it, thank you!

**Reply by klingaard** [reply]
>  I did not destroy the factory and treenode explicitly, Only in my top model, I use the getRoot()->enterTeardown() to free the resource automatically in my Destructor function.

So you follow the pattern in the examples?  Do you use the ResourceSet to maintain the factories for you?  For example: https://github.com/sparcians/map/blob/64e119f0321ca144b79071592513baedf028f958/sparta/example/SkeletonPipeline/src/SkeletonSimulator.cpp#L30-L31

> by the way, has the future presentation mentioned in the last slice[overview presentation] been done? 

Not sure what you're asking -- are you asking if there is a recording of the presentation somewhere?

**Reply by shibin2018** [reply]
1. I see, I will follow the pattern in the examples
2. Is there more presentations about the sparta besides the [overview presentation](https://docs.google.com/presentation/d/e/2PACX-1vS1BWtVv0x3qXKQWAeECe2gsF9cMG3Zp2HnXJw52grCAcl21lv3a9pLW6J0lZ32e5DWdZkFyUMcE_AI/pub?start=false&loop=false&delayms=3000&slide=id.g121c57dca98_0_41)

**Reply by klingaard** [reply]
There isn't another presentation besides the overview right now.  Most of the documentation for Sparta is in [Doxygen](https://sparcians.github.io/map/index.html) but I'm thinking about starting a full blown documentation project that reads like a big "how-to" for using the framework.  It's a big undertaking, however.
